File: scm-java/src/main/scheme/primitives/PrimitiveLoadModule.java
package scheme.primitives;

import java.io.File;
import java.io.InputStream;
import java.nio.charset.StandardCharsets;

import scheme.ModulePath;
import scheme.ModuleSource;
import scheme.Modules;
import scheme.Pair;
import scheme.Primitive;
import scheme.Scheme;
import scheme.SchemeError;
import scheme.SourcePos;
import scheme.Value;

public class PrimitiveLoadModule extends Primitive {
    private Modules modules;

    public PrimitiveLoadModule(Modules modules) {
        this.modules = modules;
    }

    @Override
    public String name() {
        return "%load-module";
    }

    @Override
    public String info() {
        return "Syntax: (%load-module module-name)\n" +
               "Library: (scm core)\n" +
               "Description: Internal primitive. Loads the named library module from an embedded .sld file or from the module search path. Returns #t if loaded successfully.\n" +
               "Example:\n" +
               "  (%load-module '(scheme base))";
    }

    @Override
    public Object apply(SourcePos pos, Object[] arguments) {
        checkArgs(pos, arguments, 1, 2);
        var moduleName = Modules.asModuleName(arguments[0]);
        if (modules.hasModule(moduleName))
            return Value.T;
        if (modules.isLoading(moduleName)) {
            throw new SchemeError(pos,
                "import: circular dependency detected while loading '~a'", moduleName);
        }

        modules.markLoading(moduleName);
        var originalModule = modules.getCurrentModule();
        try {
            var scheme = new Scheme(modules);

            // 1. Check embedded built-in .sld first
            InputStream sldStream = ModulePath.findBuiltinLibraryStream(moduleName, ".sld");
            if (sldStream != null) {
                modules.setCurrentModule(moduleName);
                var text = new String(sldStream.readAllBytes(), StandardCharsets.UTF_8);
                scheme.evalString(text, moduleName + ".sld");
                modules.updateModuleVar();
                return Value.T;
            }

            // 2. Check filesystem and .slz archives
            var source = ModulePath.findModule(modules, moduleName);
            if (source != null) {
                modules.setModuleLoadPath(moduleName, source.getDirectory());
                var scmCore = modules.getModuleRequired(null, "scm core");
                var oldSearchPath = scmCore.resolve(null, "*module-search-path*");
                try {
                    scmCore.bind("*module-search-path*", new Pair(source.getDirectory().toCharArray(), oldSearchPath));
                    modules.setCurrentModule(moduleName);
                    if (source.getContent() != null)
                        scheme.evalString(source.getContent(), source.getFileName());
                    else
                        scheme.evalFile(new File(source.getFilePath()));
                    modules.updateModuleVar();
                    return Value.T;
                } finally {
                    scmCore.bind("*module-search-path*", oldSearchPath);
                }
            }

            throw new SchemeError(pos, "Modules: " + moduleName + " not available");
        } catch (SchemeError e) {
            throw e;
        } catch (Exception e) {
            throw new SchemeError(pos, name() + ": io failure");
        } finally {
            modules.unmarkLoading(moduleName);
            modules.setCurrentModule(originalModule.getName());
        }
    }
}

File: scm-csharp/scheme/primitives/PrimitiveLoadModule.cs
using System.Text;

namespace scheme;

public class PrimitiveLoadModule : Primitive
{
    private Modules modules;

    public PrimitiveLoadModule(Modules modules)
    {
        this.modules = modules;
    }

    public override string Name() => "%load-module";

    public override string Info() =>
        "Syntax: (%load-module module-name)\n" +
        "Library: (scm core)\n" +
        "Description: Internal primitive. Loads the named library module from an embedded .sld file or from the module search path. Returns #t if loaded successfully.\n" +
        "Example:\n" +
        "  (%load-module '(scheme base))";

    public override object Apply(SourcePos? pos, object[] arguments)
    {
        CheckArgs(pos, arguments, 1, 2);
        var moduleName = Modules.AsModuleName(arguments[0]);
        // Serialize library loading so concurrent imports don't race on
        // the modules dict and per-module bindings. Reentrant lock allows
        // transitive (%load-module) calls from inside a load.
        lock (modules.LoadLock)
        {
        if (modules.HasModule(moduleName))
        {
            return Value.T;
        }
        if (modules.IsLoading(moduleName))
        {
            throw new SchemeError(pos,
                "import: circular dependency detected while loading '~a'", moduleName);
        }

        modules.MarkLoading(moduleName);
        var originalModule = modules.GetCurrentModule();
        try
        {
            var scheme = new Scheme(modules);

            // 1. Check embedded built-in .sld first
            var sldStream = ModulePath.FindBuiltinLibraryStream(moduleName, ".sld");
            if (sldStream != null)
            {
                modules.SetCurrentModule(moduleName);
                var text = new StreamReader(sldStream, Encoding.UTF8).ReadToEnd();
                scheme.EvalString(text, moduleName + ".sld");
                modules.UpdateModuleVar();
                return Value.T;
            }

            // 2. Check filesystem and .slz archives
            var source = ModulePath.FindModule(modules, moduleName);
            if (source != null)
            {
                modules.SetModuleLoadPath(moduleName, source.Directory);
                var scmCore = modules.GetModuleRequired(null, "scm core");
                var oldSearchPath = scmCore.Resolve(null, "*module-search-path*");
                try
                {
                    scmCore.Bind("*module-search-path*", new Pair(source.Directory.ToCharArray(), oldSearchPath));
                    modules.SetCurrentModule(moduleName);
                    if (source.Content != null)
                        scheme.EvalString(source.Content, source.FileName);
                    else
                        scheme.EvalFile(source.FilePath!);
                    modules.UpdateModuleVar();
                    return Value.T;
                }
                finally
                {
                    scmCore.Bind("*module-search-path*", oldSearchPath);
                }
            }

            throw new SchemeError(pos, "Modules: " + moduleName + " not available");
        }
        catch (SchemeError)
        {
            throw;
        }
        catch (Exception)
        {
            throw new SchemeError(pos, Name() + ": io failure");
        }
        finally
        {
            modules.UnmarkLoading(moduleName);
            modules.SetCurrentModule(originalModule.Name);
        }
        } // lock (modules.LoadLock)
    }
}
